File: src/validators/tools.ts
import { ValidationResult, ValidationIssue } from './index.js';
import { executeCommand } from '../utils/executor.js';

const TOOL_INFO: Record<string, { installCmd: string; description: string }> = {
  fastlane: {
    installCmd: 'gem install fastlane',
    description: 'Fastlane automation toolkit',
  },
  xcodebuild: {
    installCmd: 'xcode-select --install',
    description: 'Xcode command line tools',
  },
  gradle: {
    installCmd: 'Install Android Studio or use ./gradlew wrapper',
    description: 'Gradle build tool',
  },
  firebase: {
    installCmd: 'npm install -g firebase-tools',
    description: 'Firebase CLI',
  },
  appcenter: {
    installCmd: 'npm install -g appcenter-cli',
    description: 'AppCenter CLI',
  },
};

export async function validateTools(requiredTools: string[]): Promise<ValidationResult> {
  const issues: ValidationIssue[] = [];

  for (const tool of requiredTools) {
    const isAvailable = await checkToolAvailable(tool);

    if (!isAvailable) {
      const info = TOOL_INFO[tool] || { installCmd: `Install ${tool}`, description: tool };
      issues.push({
        level: 'error',
        code: 'TOOL_NOT_FOUND',
        message: `${info.description} (${tool}) is not installed or not in PATH`,
        suggestion: `Install with: ${info.installCmd}`,
      });
    }
  }

  return {
    valid: !issues.some(i => i.level === 'error'),
    issues,
  };
}

async function checkToolAvailable(tool: string): Promise<boolean> {
  try {
    const result = await executeCommand('which', [tool], { timeout: 5000 });
    return result.exitCode === 0;
  } catch {
    return false;
  }
}

File: src/fastlane_mcp/validators/tools.py
"""Tool availability validation."""

from fastlane_mcp.validators.types import IssueLevel, ValidationIssue
from fastlane_mcp.utils.executor import execute_command


# Installation suggestions for common tools
TOOL_INSTALL_HINTS = {
    "fastlane": "Install with: gem install fastlane OR brew install fastlane",
    "xcodebuild": "Install Xcode from the App Store",
    "gradle": "Install with: brew install gradle",
    "bundler": "Install with: gem install bundler",
    "pod": "Install with: gem install cocoapods",
    "ruby": "Install with: brew install ruby",
}


async def validate_tools(required_tools: list[str]) -> list[ValidationIssue]:
    """Validate that required tools are available.

    Args:
        required_tools: List of required tool names

    Returns:
        List of validation issues found
    """
    issues: list[ValidationIssue] = []

    for tool_name in required_tools:
        result = await execute_command("which", [tool_name])

        if result.exit_code != 0:
            suggestion = TOOL_INSTALL_HINTS.get(
                tool_name,
                f"Install {tool_name} and ensure it's in your PATH"
            )

            issues.append(ValidationIssue(
                level=IssueLevel.ERROR,
                code="TOOL_MISSING",
                message=f"Required tool not found: {tool_name}",
                suggestion=suggestion
            ))

    return issues
